worker/src/services/IngestionService/index.ts

```diff
@@ -0,0 +1,788 @@
+
+    const finalOutputCost =
+      tokenCounts.output_usage_units != null && model?.outputPrice
+        ? model.outputPrice.toNumber() * tokenCounts.output_usage_units
+        : finalInputCost
+          ? 0
+          : undefined;
+
+    const finalTotalCost =
+      tokenCounts.total_usage_units != null && model?.totalPrice
+        ? model.totalPrice.toNumber() * tokenCounts.total_usage_units
+        : finalInputCost ?? finalOutputCost
+          ? (finalInputCost ?? 0) + (finalOutputCost ?? 0)
+          : undefined;
+
+    return {
+      input_cost: finalInputCost,
+      output_cost: finalOutputCost,
+      total_cost: finalTotalCost,
+    };
+  }
+
+  private async getClickhouseRecord(params: {
+    projectId: string;
+    entityId: string;
+    table: TableName.Traces;
+  }): Promise<TraceRecordInsertType | null>;
+  private async getClickhouseRecord(params: {
+    projectId: string;
+    entityId: string;
+    table: TableName.Scores;
+  }): Promise<ScoreRecordInsertType | null>;
+  private async getClickhouseRecord(params: {
+    projectId: string;
+    entityId: string;
+    table: TableName.Observations;
+  }): Promise<ObservationRecordInsertType | null>;
+  private async getClickhouseRecord(params: {
+    projectId: string;
+    entityId: string;
+    table: TableName;
+  }) {
+    const recordParser = {
+      traces: traceRecordReadSchema,
+      scores: scoreRecordReadSchema,
+      observations: observationRecordReadSchema,
+    };
+    const { projectId, entityId, table } = params;
+
+    return await instrumentAsync({ name: `get-${table}` }, async () => {
+      const queryResult = await clickhouseClient.query({
+        query: `SELECT * FROM ${table} FINAL where project_id = '${projectId}' and id = '${entityId}'`,
+        format: "JSONEachRow",
+      });
+
+      const result = await queryResult.json();
+
+      if (result.length === 0) return null;
+
+      return table === TableName.Traces
+        ? convertTraceReadToInsert(recordParser[table].parse(result[0]))
+        : table === TableName.Scores
+          ? convertScoreReadToInsert(recordParser[table].parse(result[0]))
+          : convertObservationReadToInsert(
+              recordParser[table].parse(result[0])
+            );
+    });
+  }
+
+  private mapTraceEventsToRecords(params: {
+    traceEventList: TraceEventType[];
+    projectId: string;
+    entityId: string;
+  }): TraceRecordInsertType[] {
+    const { traceEventList, projectId, entityId } = params;
+
+    return traceEventList.map((trace) => {
+      const traceRecord: TraceRecordInsertType = {
+        id: entityId,
+        // in the default implementation, we set timestamps server side if not provided.
+        // we need to insert timestamps here and change the SDKs to send timestamps client side.
+        timestamp: this.getMicrosecondTimestamp(trace.body.timestamp),
+        name: trace.body.name,
+        user_id: trace.body.userId,
+        metadata: trace.body.metadata
```

Comment:
we need to apply the same logic as for observations below.

Reply:
// metadata needs to be converted to a record<string, string>.
      // prefix all keys with "metadata." if they are an array or primitive